File: Interview Preparation Kit/Dictionaries and Hashmaps/FrequencyQueries.js
// https://www.hackerrank.com/challenges/frequency-queries/problem
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');

    main();
});

function readLine() {
    return inputString[currentLine++];
}

// Complete the freqQuery function below.
function freqQuery(queries) {
    let output = [];
    let map = new Map();
    let frequencies = new Map();
    
    for (let i = 0; i < queries.length; i++) {
        let queryType = queries[i][0];
        let data = queries[i][1];
        
        if (queryType == 1) {
            if (map.has(data)) {
                let old_frequency_key = map.get(data);
                let new_frequency_key = old_frequency_key + 1;
                map.set(data, new_frequency_key);
                
                let old_frequency_value = frequencies.get(old_frequency_key);
                frequencies.set(old_frequency_key, old_frequency_value - 1);

                if (frequencies.has(new_frequency_key)) {
                    let new_frequency_value = frequencies.get(new_frequency_key);
                    frequencies.set(new_frequency_key, new_frequency_value + 1);
                }
                else {
                    frequencies.set(new_frequency_key, 1);
                }
            }
            else {
                map.set(data, 1);
                if (frequencies.has(1)) {
                    frequencies.set(1, frequencies.get(1) + 1);
                }
                else {
                    frequencies.set(1, 1);
                }
            }
        }
        else if (queryType == 2) {
            if (map.has(data) && map.get(data) > 1) {
                let old_frequency_key = map.get(data);
                let new_frequency_key = old_frequency_key - 1;
                map.set(data, new_frequency_key);

                let old_frequency_value = frequencies.get(old_frequency_key);
                frequencies.set(old_frequency_key, old_frequency_value - 1);

                if (frequencies.has(new_frequency_key)) {
                    frequencies.set(new_frequency_key, 
                        frequencies.get(new_frequency_key) + 1);
                }
                else {
                    frequencies.set(new_frequency_key, 1);
                }
            }
            else if (map.has(data)) {
                map.delete(data);
                if (frequencies.has(1)) {
                    frequencies.set(1, frequencies.get(1) - 1);
                }
            }
        }
        else {
            if (frequencies.has(data) && frequencies.get(data) > 0) {
                output.push(1);
            }
            else {
                output.push(0);
            }
        }
    }
    
    return output;
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const q = parseInt(readLine().trim(), 10);

    let queries = Array(q);

    for (let i = 0; i < q; i++) {
        queries[i] = readLine().replace(/\s+$/g, '').split(' ').map(queriesTemp => parseInt(queriesTemp, 10));
    }

    const ans = freqQuery(queries);

    ws.write(ans.join('\n') + '\n');

    ws.end();
}

File: Interview Preparation Kit/Dictionaries and Hashmaps/FrequencyQueries.go
// https://www.hackerrank.com/challenges/frequency-queries/problem
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

// Complete the freqQuery function below.
func freqQuery(queries [][]int32) []int32 {
	output := make([]int32, 0)
	dict, frequencies := make(map[int32]int32), make(map[int32]int32)

	for _, current := range queries {
		queryType := current[0]
		data := current[1]

		if queryType == 1 {
			first_val, exists1 := dict[data]
			if exists1 {
				dict[data]++
				frequencies[first_val]--

				_, exists2 := frequencies[first_val+1]
				if exists2 {
					frequencies[first_val+1]++
				} else {
					frequencies[first_val+1] = 1
				}
			} else {
				dict[data] = 1
				_, exists3 := frequencies[1]
				if exists3 {
					frequencies[1]++
				} else {
					frequencies[1] = 1
				}
			}
		} else if queryType == 2 {
			first_val, exists1 := dict[data]
			if exists1 && first_val > 1 {
				dict[data]--
				frequencies[first_val]--

				_, exists2 := frequencies[first_val-1]
				if exists2 {
					frequencies[first_val-1]++
				} else {
					frequencies[first_val-1] = 1
				}
			} else if exists1 {
				delete(dict, data)
				_, exists3 := frequencies[1]
				if exists3 {
					frequencies[1]--
				}
			}
		} else {
			val, exists := frequencies[data]
			if exists && val > 0 {
				output = append(output, 1)
			} else {
				output = append(output, 0)
			}
		}
	}

	return output
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 16*1024*1024)

	qTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
	checkError(err)
	q := int32(qTemp)

	var queries [][]int32
	for i := 0; i < int(q); i++ {
		queriesRowTemp := strings.Split(strings.TrimRight(readLine(reader), " \t\r\n"), " ")

		var queriesRow []int32
		for _, queriesRowItem := range queriesRowTemp {
			queriesItemTemp, err := strconv.ParseInt(queriesRowItem, 10, 64)
			checkError(err)
			queriesItem := int32(queriesItemTemp)
			queriesRow = append(queriesRow, queriesItem)
		}

		if len(queriesRow) != 2 {
			panic("Bad input")
		}

		queries = append(queries, queriesRow)
	}

	ans := freqQuery(queries)

	for i, ansItem := range ans {
		fmt.Fprintf(writer, "%d", ansItem)

		if i != len(ans)-1 {
			fmt.Fprintf(writer, "\n")
		}
	}

	fmt.Fprintf(writer, "\n")

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
